media/js/base/base-page-init.js

```diff
@@ -5,22 +5,47 @@
 /**
  * General DOM ready handler applied to all pages in base template.
  */
-$(document).ready(function() {
+(function($) {
+    // page must be loaded and ready before onWindowLoad fires
+    var loaded = false;
+    var ready = false;
 
-    var client = Mozilla.Client;
-    var utils = Mozilla.Utils;
+    function onWindowLoad() {
+        $('html').addClass('loaded');
+    }
 
-    utils.initDownloadLinks();
-    utils.initMobileDownloadLinks();
-    utils.initLangSwitcher();
+    $(document).ready(function() {
 
-    /* Bug 1264843: In partner distribution of desktop Firefox, switch the
-       downloads to corresponding partner build of Firefox for Android. */
-    if (client.isFirefoxDesktop) {
-        client.getFirefoxDetails(utils.maybeSwitchToDistDownloadLinks);
-    }
+        var client = Mozilla.Client;
+        var utils = Mozilla.Utils;
+
+        utils.initDownloadLinks();
+        utils.initMobileDownloadLinks();
+        utils.initLangSwitcher();
+
+        /* Bug 1264843: In partner distribution of desktop Firefox, switch the
+        downloads to corresponding partner build of Firefox for Android. */
+        if (client.isFirefoxDesktop) {
+            client.getFirefoxDetails(utils.maybeSwitchToDistDownloadLinks);
+        }
+
+        // if window.load happened already, fire onWindowLoad
+        if (loaded) {
+            onWindowLoad();
+        }
+
+        // note that document.ready happened to inform window.load
+        ready = true;
+    });
 
     $(window).on('load', function () {
-        $('html').addClass('loaded');
+        // if document.ready happened already, fire onWindowLoad
+        if (ready) {
```

Comment:
This is changing the behavior of the loaded class that we rely on for initiating functional tests. I'm not sure we want to trigger it on DOM ready this way, since assets such as images may not have loaded yet, or JS may not have fully excecuted. The test run looks green even with this change which is good, but it may be more prone to intermittent failures this way due to slow network conditions?

Reply:
My intent was to *not* change the behavior. The current implementation waits for both DOM ready and window load, which is what I'm trying to replicate here by making sure both `ready` and `loaded` are `true` prior to adding the `loaded` class. Maybe I'm missing something obvious though?

Reply:
I'm not sure if this is perhaps over-complicating it a bit, but maybe I'm missing something? The previous behavior added a window `load` event handler inside a DOM ready handler. Given that DOM ready always before load, all that really matters here is that the loaded class gets added on the `load` event, right?

Reply:
I'm not sure ready will always fire before load in 3.0. From the [GitHub thread linked yesterday](https://github.com/jquery/jquery/issues/3194#issuecomment-228556922):

> We recently made ready handlers fire asynchronously. This has advantages that are hard to give up. The disadvantage is that the ready handler can sometimes fire after the load event if the load event fires quickly enough.

This is probably a pretty edge-y case, but if guy from jQuery core team says it can happen, I don't mind being on the safe side.

Reply:
Ah I see,

To make this a bit easier to read/follow, would the [pattern here](https://github.com/jquery/jquery/issues/3194#issuecomment-228562930) be worth considering? It seems a bit clearer and easier to follow without needing the `load` even handler nested inside a ready handler?

Reply:
Actually, ignore my last comment. We wouldn't want to wait for both events to be satisfied before excecuting everything, right? So what you have is likely fine (and tests pass) 👍